ironfish-rust/src/public_account/description.rs

```diff
@@ -0,0 +1,218 @@
+        reader.read_exact(&mut version_buf)?;
+        let version = version_buf[0];
+        
+        let mut threshold_buf = [0; 2];
+        reader.read_exact(&mut threshold_buf)?;
+        let threshold = i16::from_le_bytes(threshold_buf);
+
+        let mut signers_len_buf = [0; 2];
+        reader.read_exact(&mut signers_len_buf)?;
+        let signers_len = i16::from_le_bytes(signers_len_buf) as usize;
+
+        let mut signers = Vec::with_capacity(signers_len);
+        for _ in 0..signers_len {
+            let mut signer_buf = [0; 32];
+            reader.read_exact(&mut signer_buf)?;
+            let signer = VerifyingKey::from_bytes(&signer_buf)
+                .map_err(|_| IronfishError::new(IronfishErrorKind::InvalidData))?;
+            signers.push(signer);
+        }
+
+        let mut signatures_len_buf = [0; 2];
+        reader.read_exact(&mut signatures_len_buf)?;
+        let signatures_len = i16::from_le_bytes(signatures_len_buf) as usize;
+
+        let mut signatures = Vec::with_capacity(signatures_len);
+        for _ in 0..signatures_len {
+            let mut signature_buf = [0; 64];
+            reader.read_exact(&mut signature_buf)?;
+            let signature = Signature::from_bytes(&signature_buf);
+            signatures.push(signature);
+        }
+
+        let mut transfers_len_buf = [0; 2];
+        reader.read_exact(&mut transfers_len_buf)?;
+        let transfers_len = i16::from_le_bytes(transfers_len_buf) as usize;
+
+        let mut transfers = Vec::with_capacity(transfers_len);
+        for _ in 0..transfers_len {
+            let transfer = Transfer::read(&mut reader)?;
+            transfers.push(transfer);
+        }
+        let address = PublicAddress::read(&mut reader)?;
+
+
+        Ok(Self {
+            version,
+            threshold,
+            signers,
+            signatures,
+            address,
+            transfers,
+        })
+    }
+
+    pub fn write<W: io::Write>(&self, mut writer: W) -> Result<(), IronfishError> {
+        // TODO: think about ideal ordering here
+        writer.write_all(&self.version.to_le_bytes())?;
+        
+        writer.write_all(&self.threshold.to_le_bytes())?;
+
+        let signers_len = self.signers.len() as i16;
+        writer.write_all(&signers_len.to_le_bytes())?;
+
+        for signer in &self.signers {
+            writer.write_all(signer.as_bytes())?;
+        }
+
+        let signatures_len = self.signatures.len() as i16;
+        writer.write_all(&signatures_len.to_le_bytes())?;
+
+        for signature in &self.signatures {
+            writer.write_all(&signature.to_bytes())?;
+        }
+
+        let transfers_len = self.transfers.len() as i16;
+        writer.write_all(&transfers_len.to_le_bytes())?;
+
+        for transfer in &self.transfers {
+            transfer.write(&mut writer)?;
+        }
+
+        self.address.write(&mut writer)?;
+
+        Ok(())
+    }
+
+    pub fn valid(&self) -> Result<(), IronfishError> {
```

Comment:
We shouldn't allow to instantiate an invalid struct. This method should be private and should be automatically called by all other methods that create a new `PublicAccountDescription` (such as `read()`)